File: packages/sdk/sdk-java/src/app/phaseo/sdk/Workflows.java
package app.phaseo.sdk;

import java.time.Duration;
import java.util.Iterator;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.Set;
import java.util.function.Function;

public final class Workflows {
    private Workflows() {}

    public record Page<T>(List<T> data, boolean hasMore) {}

    @FunctionalInterface
    public interface PageFetcher<T> { Page<T> fetch(int limit, int offset); }

    public static final class PageIterator<T> implements Iterator<Page<T>>, Iterable<Page<T>> {
        private final PageFetcher<T> fetcher;
        private final int limit;
        private int offset;
        private boolean done;

        public PageIterator(int limit, int offset, PageFetcher<T> fetcher) {
            this.limit = Math.max(1, limit);
            this.offset = Math.max(0, offset);
            this.fetcher = fetcher;
        }

        @Override public Iterator<Page<T>> iterator() { return this; }
        @Override public boolean hasNext() { return !done; }
        @Override public Page<T> next() {
            if (done) throw new NoSuchElementException();
            Page<T> page = fetcher.fetch(limit, offset);
            int count = page.data() == null ? 0 : page.data().size();
            offset += count;
            done = !page.hasMore() || count == 0;
            return page;
        }
    }

    public static final class JobHandle<T> {
        private static final Set<String> FAILED = Set.of("failed", "cancelled", "canceled", "expired");
        private final String kind;
        private final String id;
        private final Function<String, T> fetcher;
        private final Function<T, String> status;

        public JobHandle(String kind, String id, Function<String, T> fetcher, Function<T, String> status) {
            if (id == null || id.isBlank()) throw new IllegalArgumentException("Job ID is required");
            this.kind = kind;
            this.id = id;
            this.fetcher = fetcher;
            this.status = status;
        }

        public String kind() { return kind; }
        public String id() { return id; }
        public T refresh() { return fetcher.apply(id); }

        public T waitForCompletion(Duration interval) throws InterruptedException {
            long delay = Math.max(1, interval.toMillis());
            while (true) {
                T value = refresh();
                String current = status.apply(value).toLowerCase();
                if (current.equals("completed")) return value;
                if (FAILED.contains(current)) throw new IllegalStateException(kind + " job " + id + " ended with status " + current);
                Thread.sleep(delay);
            }
        }
    }
}

File: packages/sdk/sdk-rust/src/workflows.rs
use std::thread;
use std::time::Duration;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Page<T> {
    pub data: Vec<T>,
    pub has_more: bool,
}

pub struct Pager<T, E, F>
where
    F: FnMut(usize, usize) -> Result<Page<T>, E>,
{
    fetch: F,
    limit: usize,
    offset: usize,
    done: bool,
}

impl<T, E, F> Pager<T, E, F>
where
    F: FnMut(usize, usize) -> Result<Page<T>, E>,
{
    pub fn new(limit: usize, offset: usize, fetch: F) -> Self {
        Self {
            fetch,
            limit: limit.max(1),
            offset,
            done: false,
        }
    }
}

impl<T, E, F> Iterator for Pager<T, E, F>
where
    F: FnMut(usize, usize) -> Result<Page<T>, E>,
{
    type Item = Result<Page<T>, E>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.done {
            return None;
        }
        match (self.fetch)(self.limit, self.offset) {
            Ok(page) => {
                let count = page.data.len();
                self.offset += count;
                self.done = !page.has_more || count == 0;
                Some(Ok(page))
            }
            Err(error) => {
                self.done = true;
                Some(Err(error))
            }
        }
    }
}

pub struct JobHandle<T, E, F, S>
where
    F: FnMut(&str) -> Result<T, E>,
    S: for<'a> Fn(&'a T) -> &'a str,
{
    pub kind: String,
    pub id: String,
    fetch: F,
    status: S,
}

impl<T, E, F, S> JobHandle<T, E, F, S>
where
    F: FnMut(&str) -> Result<T, E>,
    S: for<'a> Fn(&'a T) -> &'a str,
{
    pub fn new(kind: impl Into<String>, id: impl Into<String>, fetch: F, status: S) -> Self {
        Self {
            kind: kind.into(),
            id: id.into(),
            fetch,
            status,
        }
    }

    pub fn refresh(&mut self) -> Result<T, E> {
        (self.fetch)(&self.id)
    }

    pub fn wait(&mut self, interval: Duration) -> Result<T, String>
    where
        E: std::fmt::Display,
    {
        loop {
            let value = self.refresh().map_err(|error| error.to_string())?;
            let current = (self.status)(&value).to_ascii_lowercase();
            if current == "completed" {
                return Ok(value);
            }
            if matches!(
                current.as_str(),
                "failed" | "cancelled" | "canceled" | "expired"
            ) {
                return Err(format!(
                    "{} job {} ended with status {}",
                    self.kind, self.id, current
                ));
            }
            thread::sleep(interval);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pagination_and_job_handles_advance_safely() {
        let mut pages = Pager::new(2, 0, |_, offset| -> Result<Page<i32>, String> {
            Ok(if offset == 0 {
                Page {
                    data: vec![1, 2],
                    has_more: true,
                }
            } else {
                Page {
                    data: vec![3],
                    has_more: false,
                }
            })
        });
        assert_eq!(pages.next().unwrap().unwrap().data, vec![1, 2]);
        assert_eq!(pages.next().unwrap().unwrap().data, vec![3]);
        assert!(pages.next().is_none());

        let mut statuses = vec!["completed".to_string(), "running".to_string()];
        let mut job = JobHandle::new(
            "video",
            "video_1",
            move |_| -> Result<String, String> { Ok(statuses.pop().unwrap()) },
            |value: &String| value.as_str(),
        );
        assert_eq!(job.wait(Duration::from_millis(1)).unwrap(), "completed");
    }
}
